include/sparrow/buffer_adaptor.hpp

```diff
@@ -0,0 +1,499 @@
+// Copyright 2024 Man Group Operations Limited
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cmath>
+#include <limits>
+#include <utility>
+
+#include "sparrow/buffer.hpp"
+#include "sparrow/contracts.hpp"
+
+namespace sparrow
+{
+    /**
+     * @brief Class which have internally a reference to a buffer<T> and provides the API of a buffer<U>
+     */
+    template <typename T, typename U>
+        requires(sizeof(U) < sizeof(T))
```

Comment:
Why this requirement?

Reply:
If sizeof(U) > sizeof(T), for example U is 8 bytes and T is 1 bytes.
If you want to insert/resize 2 T elements, you have to insert 1 U element.
So after inserting 2 T elements, the size of the buffer_adaptor is now 8. 
BTW, we should also have the requirement that sizeof(T) i a multiple of sizeof(U).

Reply:
Indeed, but I wonder if we should not add some constraints on the number of new elements we can insert / push_back etc in that case. That goes beyond our use case though (we will mostly convert from uint8_t to anything else), so let's keep it as is for now.